presto-main/src/main/java/com/facebook/presto/execution/buffer/ArbitraryOutputBuffer.java

```diff
@@ -248,11 +253,20 @@ public void enqueue(Lifespan lifespan, List<SerializedPage> pages)
         masterBuffer.addPages(serializedPageReferences);
 
         // process any pending reads from the client buffers
-        for (ClientBuffer clientBuffer : safeGetBuffersSnapshot()) {
+        List<ClientBuffer> buffers = safeGetBuffersSnapshot();
+        if (buffers.isEmpty()) {
+            return;
+        }
+        // handle potential for racy update of next index and client buffers present
+        int index = nextClientBufferIndex.get() % buffers.size();
+        for (int i = 0; i < buffers.size(); i++) {
+            buffers.get(index).loadPagesIfNecessary(masterBuffer);
+            index = (index + 1) % buffers.size();
             if (masterBuffer.isEmpty()) {
+                // Resume from the next client buffer on the next iteration
+                nextClientBufferIndex.set(index);
```

Comment:
Is enqueue called from multiple threads? IOW do multiple threads enter this loop and fetch nextClientBufferIndex?

Reply:
I believe it’s possible that yes, they could. I didn’t look into it because I propose that scenario is not a “problem” worth solving because if that does happen this implementation will put one thread “in front” of the other in performing the iteration at all times (even if the leading caller changes between iterations). The net result is that the leading caller releases any blocked clients and the trailing caller has no effect. That seems preferable to putting more synchronization around enqueue operations, because this path only matters when the producer side is slower than consumers. A faster producer still drives the output distribution through consumption rate.

Reply:
WFM